procps/uptime.rs

```diff
@@ -1,67 +1,39 @@
 use libc;
-use libc::printf;
+use libc::getutxent;
+use libc::localtime;
+use libc::sysinfo;
 use libc::time;
-extern "C" {
 
-  #[no_mangle]
-  fn localtime(__timer: *const time_t) -> *mut tm;
-  #[no_mangle]
-  fn getutxent() -> *mut utmpx;
+extern "C" {
   #[no_mangle]
   fn getopt32(argv: *mut *mut libc::c_char, applet_opts: *const libc::c_char, _: ...) -> u32;
-  #[no_mangle]
-  fn sysinfo(__info: *mut sysinfo) -> libc::c_int;
 }
 
-use libc::time_t;
-use libc::tm;
-#[derive(Copy, Clone)]
-#[repr(C)]
-pub struct __exit_status {
-  pub e_termination: libc::c_short,
-  pub e_exit: libc::c_short,
-}
-use libc::utmpx;
-#[derive(Copy, Clone)]
-#[repr(C)]
-pub struct C2RustUnnamed {
-  pub tv_sec: i32,
-  pub tv_usec: i32,
-}
-pub type __u16 = libc::c_ushort;
-pub type u32 = libc::c_uint;
-pub type __kernel_long_t = libc::c_long;
-pub type __kernel_ulong_t = libc::c_ulong;
-#[derive(Copy, Clone)]
-#[repr(C)]
-pub struct sysinfo {
-  pub uptime: __kernel_long_t,
-  pub loads: [__kernel_ulong_t; 3],
-  pub totalram: __kernel_ulong_t,
-  pub freeram: __kernel_ulong_t,
-  pub sharedram: __kernel_ulong_t,
-  pub bufferram: __kernel_ulong_t,
-  pub totalswap: __kernel_ulong_t,
-  pub freeswap: __kernel_ulong_t,
-  pub procs: __u16,
-  pub pad: __u16,
-  pub totalhigh: __kernel_ulong_t,
-  pub freehigh: __kernel_ulong_t,
-  pub mem_unit: u32,
-  pub _f: [libc::c_char; 0],
+// When reading the utmp entries with getuxent, this identifies
+// a user entry.
+const UTMP_USER_PROCESS: libc::c_short = 7;
+
+fn get_users() -> u32 {
+  let mut users = 0;
+  unsafe {
+    while let Some(ut) = getutxent().as_mut() {
+      if ut.ut_type == UTMP_USER_PROCESS && ut.ut_user[0] != 0 {
+        users += 1
+      }
+    }
+  }
+  users
 }
+
 /* nr of bits of precision */
 /* 1.0 as fixed-point */
 #[no_mangle]
-pub unsafe extern "C" fn uptime_main(
-  mut _argc: libc::c_int,
-  mut argv: *mut *mut libc::c_char,
-) -> libc::c_int {
-  let mut updays: libc::c_uint = 0;
-  let mut uphours: libc::c_uint = 0;
-  let mut upminutes: libc::c_uint = 0;
-  let mut opts: libc::c_uint = 0;
-  let mut info: sysinfo = sysinfo {
+pub extern "C" fn uptime_main(mut _argc: libc::c_int, argv: *mut *mut libc::c_char) -> libc::c_int {
+  let opts = unsafe { getopt32(argv, b"s\x00" as *const u8 as *const libc::c_char) };
+
+  let mut current_secs = unsafe { time(std::ptr::null_mut()) };
+
+  let mut info = sysinfo {
```

Comment:
Let's use https://doc.rust-lang.org/std/mem/fn.zeroed.html here instead of writing down all the fields.

Reply:
Oh thanks, I'm still very new to unsafe Rust.